posts/elixir-graphql-part-one.md

```diff
@@ -0,0 +1,311 @@
+        inserted_at
+        updated_at
+    }
+}
+```
+
+To make this work, we'll need to define both the Types and the Schemas for GraphQL to understand how to turn data in the database, and then data in Elixir data structures, into GraphQL data structures. We'll start off by defining the types. Create a new file, under `lib/greenfy_web/schema`, called `data_types.ex` (you'll need to create the `schema` directory), and we'll start defining our first object type:
+
+```
+defmodule Greenfy.Schema.DataTypes do
+  use Absinthe.Schema.Notation
+
+  object :event do
+    field :id, :id
+    field :event_type, :string
+    field :message, :string
+    field :payload, :string
+  end
+end
+```
+
+You'll probably notice that `:inserted_at` and `:updated_at` are not being included in our data types notation yet! The reason for this is that those are not easily represented by the standard built-in types for Absinthe, so we'll need to define some fancy `scalar` types later! (Don't worry, we'll describe this all in great detail later!)
+
+#### Sidebar: GraphQL Definitions
+
+We're using A LOT of terms here pretty liberally when talking about GraphQL but not doing a lot to explain what they actually are! Let's spend a really quick sidebar talking about some of the terms we've discussed so far (and the ones we'll be doing later).
+
+First off, we have **Types**. Types are exactly what they sound like: definitions of the various data types we'll be defining for our GraphQL application. A good example of this is the object definition above: An **event** is a GraphQL **object** type, with four **fields** defined on it so far (we'll get to fields next)!
+
+A **Field** is essentially a bit of queryable information. An object, like the data type we defined previously, stores a collection of different fields together. Each field should define, at a minimum, its name (or key) and its type!
+
+Next, we'll be working with **Schemas**. If **Types** define what data is queryable in our GraphQL API and the shape of that data, the **Schema** defines how we actually get that data OUT of our GraphQL API!
+
+Later on, you'll see something called **Resolvers**. If the **Schema** tells the client how to query the data out of our application, the **Resolver** tells the server how to react to that query and how to interpret those GraphQL queries.
+
+Finally, this will be something we'll use much later, but you'll hear it mentioned before that point: **Mutations**! Mutations are how the data in our GraphQL application is modified! Instead of your `POST`, `PUT`, and `PATCH` endpoints in your REST API, you just define **Mutations** instead!
+
+#### Back to our implementation
+
+Next, we'll need to create our `Schema`. Create `lib/greenfy_web/schema.ex` and give it the following contents:
+
+```
+defmodule Greenfy.Schema do
+  use Absinthe.Schema
+
+  import_types Greenfy.Schema.DataTypes
+
+  query do
+    @desc "Get a list of events"
+    field :events, list_of(:event) do
```

Comment:
This should probably be `list_of(non_null(:event))`.

Reply:
Yeah, I think that'll be a nice point to explore in a later post around modifiers to field descriptions!